File: pqc-algorithms/csidh/csidh_benchmark.go
package main

import (
	"bufio"
	"crypto/rand"
	"fmt"
	"math"
	"os"
	"time"

	"github.com/cloudflare/circl/dh/csidh"
)

const iterations = 10000

var out [64]byte
var rng = rand.Reader
var interruptMode = false

func main() {
	if len(os.Args) > 1 && os.Args[1] == "interrupt" {
		interruptMode = true
	}

	var keygenTimes, pubkeyTimes, sharedTimes []float64

	for i := 0; i < iterations; i++ {
		if interruptMode {
			waitForContinue("Press 'c' to continue to Key Generation")
		}
		start := time.Now()
		priv := csidh.PrivateKey{}
		csidh.GeneratePrivateKey(&priv, rng)
		elapsed := time.Since(start).Seconds() * 1000
		keygenTimes = append(keygenTimes, elapsed)

		if interruptMode {
			waitForContinue("Press 'c' to continue to Public Key Computation")
		}
		start = time.Now()
		pub := csidh.PublicKey{}
		csidh.GeneratePublicKey(&pub, &priv, rng)
		elapsed = time.Since(start).Seconds() * 1000
		pubkeyTimes = append(pubkeyTimes, elapsed)

		if interruptMode {
			waitForContinue("Press 'c' to continue to Shared Secret Derivation")
		}
		priv2 := csidh.PrivateKey{}
		csidh.GeneratePrivateKey(&priv2, rng)
		pub2 := csidh.PublicKey{}
		csidh.GeneratePublicKey(&pub2, &priv2, rng)

		start = time.Now()
		csidh.DeriveSecret(&out, &pub, &priv2, rng)
		elapsed = time.Since(start).Seconds() * 1000
		sharedTimes = append(sharedTimes, elapsed)
	}

	printStats("Key Generation", keygenTimes)
	printStats("Public Key Computation", pubkeyTimes)
	printStats("Shared Secret Derivation", sharedTimes)
}

func printStats(name string, times []float64) {
	sum := 0.0
	for _, t := range times {
		sum += t
	}
	mean := sum / float64(len(times))

	variance := 0.0
	for _, t := range times {
		variance += math.Pow(t-mean, 2)
	}
	stddev := math.Sqrt(variance / float64(len(times)))

	fmt.Printf("%s:\n", name)
	fmt.Printf("  Average Time: %.3f ms\n", mean)
	fmt.Printf("  Std Dev:      %.3f ms\n\n", stddev)
}

func waitForContinue(prompt string) {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Println(prompt)
		input, _ := reader.ReadString('\n')
		if input == "c\n" || input == "c\r\n" {
			break
		}
	}
}

File: pqc-algorithms/csidh-20210627/csidh_benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>
#include <assert.h>

#include "fp.h"
#include "csidh.h"

#define ITERATIONS 10000

static double times_keygen[ITERATIONS];
static double times_pubkey[ITERATIONS];
static double times_shared[ITERATIONS];

static int interrupt_mode = 0;

static void record_time(double *arr, int idx, clock_t start, clock_t end) {
    arr[idx] = 1000.0 * (end - start) / CLOCKS_PER_SEC;
}

static void print_stats(const char *label, double *arr, int n) {
    double sum = 0, mean = 0, stddev = 0;

    for (int i = 0; i < n; ++i)
        sum += arr[i];
    mean = sum / n;

    for (int i = 0; i < n; ++i)
        stddev += (arr[i] - mean) * (arr[i] - mean);
    stddev = sqrt(stddev / n);

    printf("%s:\n", label);
    printf("  Average Time: %.3f ms\n", mean);
    printf("  Std Dev:      %.3f ms\n\n", stddev);
}

static void wait_for_continue(const char *msg) {
    if (!interrupt_mode) return;
    printf("%s\n", msg);
    fflush(stdout);

    int c;
    do {
        c = getchar();
    } while (c != 'c' && c != 'C');
}

int main(int argc, char **argv) {
    if (argc > 1 && strcmp(argv[1], "interrupt") == 0) {
        interrupt_mode = 1;
    }

    private_key priv1, priv2;
    public_key pub1, pub2, shared;

    for (int i = 0; i < ITERATIONS; ++i) {
        clock_t start, end;

        wait_for_continue("Press 'c' to continue to Key Generation");
        start = clock();
        csidh_private(&priv1);
        end = clock();
        record_time(times_keygen, i, start, end);

        wait_for_continue("Press 'c' to continue to Public Key Computation");
        start = clock();
        csidh(&pub1, &base, &priv1);
        assert(ret);
        end = clock();
        record_time(times_pubkey, i, start, end);

        csidh_private(&priv2);
        csidh(&pub2, &base, &priv2);
        assert(ret);

        wait_for_continue("Press 'c' to continue to Shared Secret Derivation");
        start = clock();
        csidh(&shared, &pub1, &priv2);
        assert(ret);
        end = clock();
        record_time(times_shared, i, start, end);
    }

    print_stats("Key Generation", times_keygen, ITERATIONS);
    print_stats("Public Key Computation", times_pubkey, ITERATIONS);
    print_stats("Shared Secret Derivation", times_shared, ITERATIONS);

    return 0;
}

